src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -354,6 +360,156 @@ citus_relation_size(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * citus_shard_property_set allows setting shard properties for all
+ * the shards within the shard group that given shard belongs to.
+ */
+Datum
+citus_shard_property_set(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	PG_ENSURE_ARGNOTNULL(0, "shard_id");
+	uint64 shardId = PG_GETARG_INT64(0);
+
+	if (!ShardExists(shardId))
+	{
+		ereport(ERROR, (errmsg("shard %lu does not exist", shardId)));
+	}
+
+	Oid distributedRelationId = RelationIdForShard(shardId);
+	List *colocatedTableList = ColocatedTableList(distributedRelationId);
+	EnsureTableListOwner(colocatedTableList);
+
+	AcquirePlacementColocationLock(distributedRelationId, ExclusiveLock,
+								   "set anti affinity property for a shard of");
+
+	Oid colocatedTableId = InvalidOid;
+	foreach_oid(colocatedTableId, colocatedTableList)
+	{
+		LockRelationOid(colocatedTableId, ShareUpdateExclusiveLock);
+	}
+
+	if (!PG_ARGISNULL(1))
+	{
+		bool antiAffinity = PG_GETARG_BOOL(1);
+		citus_shard_property_set_anti_affinity(shardId, antiAffinity);
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_shard_property_set_anti_affinity is an helper function for
+ * citus_shard_property_set UDF to set anti_affinity property for given
+ * shard.
+ */
+static void
+citus_shard_property_set_anti_affinity(uint64 shardId, bool enabled)
+{
+	Oid distributedRelationId = RelationIdForShard(shardId);
+	if (!IsCitusTableType(distributedRelationId, HASH_DISTRIBUTED) &&
+		!IsCitusTableType(distributedRelationId, SINGLE_SHARD_DISTRIBUTED))
+	{
+		ereport(ERROR, (errmsg("shard isolation is only supported for hash "
+							   "distributed tables")));
+	}
+
+	ShardGroupSetNeedsSeparateNodeGlobally(shardId, enabled);
+}
+
+
+/*
+ * ShardGroupSetNeedsSeparateNodeGlobally calls ShardGroupSetNeedsSeparateNode
+ * on all nodes.
+ */
+static void
+ShardGroupSetNeedsSeparateNodeGlobally(uint64 shardId, bool enabled)
+{
+	ShardGroupSetNeedsSeparateNode(shardId, enabled);
+
+	char *metadataCommand =
+		ShardGroupSetNeedsSeparateNodeCommand(shardId, enabled);
+	SendCommandToWorkersWithMetadata(metadataCommand);
+}
+
+
+/*
+ * ShardGroupSetNeedsSeparateNode sets the needsseparatenode flag to desired
+ * value for all the shards within the shard group that given shard belongs
+ * to.
+ */
+void
+ShardGroupSetNeedsSeparateNode(uint64 shardId, bool enabled)
+{
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	List *colocatedShardIntervalList = ColocatedShardIntervalList(shardInterval);
+
+	ShardInterval *colocatedShardInterval = NULL;
+	foreach_ptr(colocatedShardInterval, colocatedShardIntervalList)
+	{
+		ShardSetNeedsSeparateNode(colocatedShardInterval->shardId,
+								  enabled);
+	}
```

Comment:
Say, even if we follow the first option, CitusInvalidateRelcacheByShardId() is designed to send invalidations for a single shard / would open pg_dist_shard once for each shard, hence we don't seem to gain much. What do you think? Do you think CitusInvalidateRelcacheByShardId() needs refactoring too so that it can perform the invalidations for a list of shards as well?

Reply:
Agree. Even more, realized that calling CitusInvalidateRelcacheByShardId() once for a shard group is fine as it sends the invalidation based on distributed rel id, not the shard id.

```c
CitusInvalidateRelcacheByRelid(shardForm->logicalrelid);
```

Reply:
I realized that we don't even need to call CitusInvalidateRelcacheByShardId multiple times as it anyway invalidates the caches for the distributed table that owns the shard, not just for the input shard, addressing this in next commit.